File: latticeai/core/file_generation/repair.py
"""Deterministic scaffolds — step 5, the last resort after the retries.

The user asked for a file, so the request must still end in a well-formed file
of the target type, never an error. Nothing here calls a model: whatever the
model did produce is salvaged where it can be, and honestly labelled where it
cannot.
"""

from __future__ import annotations

import ast
import html as html_lib
import json
import re

from .extraction import _ext, _slice_html_document, _slice_json_document
from .validation import looks_like_refusal


def repair_file_content(content: str, target_path: str, user_request: str) -> str:
    """Turn whatever the model produced into a valid file of the target type.

    This is the last resort after retries: the user asked for a file, so the
    request must still end in a well-formed file, never an error.
    """
    ext = _ext(target_path)
    salvage = content.strip()
    if looks_like_refusal(salvage):
        salvage = ""

    if ext in (".html", ".htm"):
        return _repair_html(salvage, user_request)
    if ext == ".json":
        sliced = _slice_json_document(salvage)
        if sliced is not None:
            return sliced
        return json.dumps(
            {"request": user_request, "content": salvage},
            ensure_ascii=False,
            indent=2,
        )
    if ext == ".py" and salvage:
        # The repair guarantee for Python is parseability: unparseable output
        # is preserved honestly as a commented-out draft, never as a broken
        # module the user has to debug.
        try:
            ast.parse(salvage)
            return salvage
        except SyntaxError:
            commented = "\n".join(f"# {line}" for line in salvage.splitlines())
            return (
                f"# TODO: model produced invalid Python for: {user_request}\n"
                "# The draft below is preserved as comments — fix and uncomment.\n"
                f"{commented}\n"
            )
    if salvage:
        return salvage
    # Nothing usable at all — leave an honest placeholder in the right format.
    comment = {
        ".py": "# TODO: model produced no usable content for: ",
        ".js": "// TODO: model produced no usable content for: ",
        ".jsx": "// TODO: model produced no usable content for: ",
        ".ts": "// TODO: model produced no usable content for: ",
        ".tsx": "// TODO: model produced no usable content for: ",
        ".css": "/* TODO: model produced no usable content for: ",
        ".sh": "# TODO: model produced no usable content for: ",
        ".sql": "-- TODO: model produced no usable content for: ",
    }.get(ext, "")
    if ext == ".css":
        return f"{comment}{user_request} */\n"
    if comment:
        return f"{comment}{user_request}\n"
    return f"{user_request}\n"


def _repair_html(salvage: str, user_request: str) -> str:
    lower = salvage.lower()
    if "<html" in lower or "<!doctype" in lower:
        # A real document that is merely truncated — close it.
        doc = _slice_html_document(salvage)
        low = doc.lower()
        if "</body>" not in low and "<body" in low:
            doc += "\n</body>"
        if "</html>" not in low:
            doc += "\n</html>"
        return doc
    if re.search(r"<\w+[^>]*>", salvage):
        body = salvage  # an HTML fragment — embed as-is
    elif salvage:
        body = "\n".join(
            f"  <p>{html_lib.escape(line)}</p>"
            for line in salvage.splitlines()
            if line.strip()
        )
    else:
        body = f"  <p>{html_lib.escape(user_request)}</p>"
    title = html_lib.escape(user_request[:60] or "Generated page")
    return (
        "<!DOCTYPE html>\n"
        "<html lang=\"ko\">\n"
        "<head>\n"
        "  <meta charset=\"utf-8\">\n"
        "  <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n"
        f"  <title>{title}</title>\n"
        "  <style>\n"
        "    body { font-family: system-ui, sans-serif; margin: 2rem auto; "
        "max-width: 720px; line-height: 1.6; padding: 0 1rem; }\n"
        "  </style>\n"
        "</head>\n"
        "<body>\n"
        f"{body}\n"
        "</body>\n"
        "</html>"
    )

File: rust/lattice-agent/src/content/sanitize/repair.rs
//! `repair.py` — deterministic scaffolds, the last resort.

use serde_json::Value;

use crate::content::pydiff::splitlines;
use crate::parse::pystr::char_slice;

use super::extract::{slice_html_document, slice_json_document};
use super::python::python_parses;
use super::text::{contains_ci, ext_of, html_tag, matched};
use super::validate::looks_like_refusal;

// ── repair (`repair.py`) ────────────────────────────────────────────────────

/// `html.escape(text)` — `quote=True`, so both quote characters go too.
fn escape_html(text: &str) -> String {
    text.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&#x27;")
}

/// `repair_file_content(content, target_path, user_request)`.
pub fn repair_file_content(content: &str, target_path: &str, user_request: &str) -> String {
    let ext = ext_of(target_path);
    let mut salvage = content.trim().to_string();
    if looks_like_refusal(&salvage) {
        salvage = String::new();
    }

    if ext == ".html" || ext == ".htm" {
        return repair_html(&salvage, user_request);
    }
    if ext == ".json" {
        if let Some(sliced) = slice_json_document(&salvage) {
            return sliced;
        }
        // `json.dumps({...}, ensure_ascii=False, indent=2)`, written out rather
        // than serialised: `serde_json`'s key order depends on whether some
        // *other* crate in the build enabled `preserve_order`, and a file's
        // bytes must not.
        let quoted = |text: &str| Value::String(text.to_string()).to_string();
        return format!(
            "{{\n  \"request\": {},\n  \"content\": {}\n}}",
            quoted(user_request),
            quoted(&salvage)
        );
    }
    if ext == ".py" && !salvage.is_empty() {
        // The repair guarantee for Python is parseability: unparseable output
        // is preserved honestly as a commented-out draft, never as a broken
        // module the user has to debug.
        if python_parses(&salvage).is_ok() {
            return salvage;
        }
        let commented = splitlines(&salvage)
            .into_iter()
            .map(|line| format!("# {line}"))
            .collect::<Vec<_>>()
            .join("\n");
        return format!(
            "# TODO: model produced invalid Python for: {user_request}\n\
# The draft below is preserved as comments — fix and uncomment.\n\
{commented}\n"
        );
    }
    if !salvage.is_empty() {
        return salvage;
    }
    // Nothing usable at all — an honest placeholder in the right format.
    let comment = match ext.as_str() {
        ".py" | ".sh" => "# TODO: model produced no usable content for: ",
        ".js" | ".jsx" | ".ts" | ".tsx" => "// TODO: model produced no usable content for: ",
        ".css" => "/* TODO: model produced no usable content for: ",
        ".sql" => "-- TODO: model produced no usable content for: ",
        _ => "",
    };
    if ext == ".css" {
        return format!("{comment}{user_request} */\n");
    }
    if !comment.is_empty() {
        return format!("{comment}{user_request}\n");
    }
    format!("{user_request}\n")
}

/// `_repair_html(salvage, user_request)`.
fn repair_html(salvage: &str, user_request: &str) -> String {
    if contains_ci(salvage, "<html") || contains_ci(salvage, "<!doctype") {
        // A real document that is merely truncated — close it.
        let mut doc = slice_html_document(salvage);
        let low = doc.to_lowercase();
        if !low.contains("</body>") && low.contains("<body") {
            doc.push_str("\n</body>");
        }
        if !low.contains("</html>") {
            doc.push_str("\n</html>");
        }
        return doc;
    }
    let body = if matched(html_tag(), salvage) {
        salvage.to_string()
    } else if !salvage.is_empty() {
        splitlines(salvage)
            .into_iter()
            .filter(|line| !line.trim().is_empty())
            .map(|line| format!("  <p>{}</p>", escape_html(line)))
            .collect::<Vec<_>>()
            .join("\n")
    } else {
        format!("  <p>{}</p>", escape_html(user_request))
    };
    let title = escape_html(if user_request.is_empty() {
        "Generated page"
    } else {
        char_slice(user_request, 60)
    });
    // Indentation is part of the file, so every line carries its own spaces
    // rather than borrowing this source file's.
    format!(
        concat!(
            "<!DOCTYPE html>\n",
            "<html lang=\"ko\">\n",
            "<head>\n",
            "  <meta charset=\"utf-8\">\n",
            "  <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n",
            "  <title>{title}</title>\n",
            "  <style>\n",
            "    body {{ font-family: system-ui, sans-serif; margin: 2rem auto; ",
            "max-width: 720px; line-height: 1.6; padding: 0 1rem; }}\n",
            "  </style>\n",
            "</head>\n",
            "<body>\n",
            "{body}\n",
            "</body>\n",
            "</html>",
        ),
        title = title,
        body = body,
    )
}
